### PR TITLE
feat(profiling): add aggregateFlamegraph to profileSummary

### PR DESCRIPTION
**Pull request by eliashussary**
## Summary

Adds aggregate flamegraphs to `profileSummary` view.

image

### CODE REVIEW

File: static/app/components/profiling/flamegraph/aggregateFlamegraph.tsx [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+  let longestFrame: FlamegraphFrame | null = null;
+
+  const frames: FlamegraphFrame[] = [...flamegraph.root.children];
+  while (frames.length > 0) {
+    const frame = frames.pop()!;
+    if (
+      focusFrame.name === frame.frame.name &&
+      // the image name on a frame is optional,
+      // treat it the same as the empty string
+      focusFrame.package === (frame.frame.image || '') &&
```
Comment (JonasBa) [inline static/app/components/profiling/flamegraph/aggregateFlamegraph.tsx:63]: Feel free to change this to be focusFrame.packge === frame.frame.image, I need to merge https://github.com/getsentry/sentry/pull/43705 which will fix this

File: static/app/components/profiling/flamegraph/aggregateFlamegraph.tsx [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+      (longestFrame?.node?.totalWeight || 0) < frame.node.totalWeight
+    ) {
+      longestFrame = frame;
+    }
+
+    for (let i = 0; i < frame.children.length; i++) {
```
Comment (JonasBa) [inline static/app/components/profiling/flamegraph/aggregateFlamegraph.tsx:69]: You can wrap this in an if(longestFrame && longestFrame.node.totalWeight < frame.totalWeight) or just skip the entire comparison step on L58 where you pop the frame off the queue by checking if(frame.node.totalWeight < longestFrame.node.totalWeight)

This way you can skip iterating over large parts of a tree as soon as you found one longestFrame because you now have a min value to compare to you know that children of a frame can never be longer than their parent

File: static/app/components/profiling/flamegraph/aggregateFlamegraph.tsx [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+    transaction.finish();
+
+    return newFlamegraph;
+  }, [profile, sorting, threadId, view]);
+
+  const flamegraphCanvas = useMemo(() => {
+    if (!flamegraphCanvasRef) {
+      return null;
+    }
+    const yOrigin =
```
Comment (JonasBa) [inline static/app/components/profiling/flamegraph/aggregateFlamegraph.tsx:137]: I shipped a change yesterday where you can remove this, it should just be `flamegraphTheme.SIZES.TIMELINE_HEIGHT * devicePixelRatio`

File: static/app/utils/profiling/flamegraph/flamegraphThemeProvider.tsx [hunk 4]
```diff
@@ -10,22 +11,47 @@ import {
 
 export const FlamegraphThemeContext = createContext<FlamegraphTheme | null>(null);
 
+type FlamegraphThemeMutationCallback = (
```
Comment (JonasBa) [inline static/app/utils/profiling/flamegraph/flamegraphThemeProvider.tsx:14]: I assume you need this because we need to mutate this from the children component right?
It smell to me like we should stop passing entire theme object to the renderers and only pass the values required by each renderer - this way we get some flexibility back and stop relying on the theme.

Comment (eliashussary) [reply static/app/utils/profiling/flamegraph/flamegraphThemeProvider.tsx:14]: Absolutely we should be passing via props.

In an ideal scenario some of our flamegraph ui components shouldn't rely so much on context. But then again, we have to weigh in on how often we're going to use a flamegraph elsewhere.  🤷🏼

File: static/app/components/profiling/flamegraph/interactions/useWheelCenterZoom.tsx [hunk 5]
```diff
@@ -9,11 +9,12 @@ import {getCenterScaleMatrixFromMousePosition} from 'sentry/utils/profiling/gl/u
...
   canvas: FlamegraphCanvas | null,
   view: CanvasView<any> | null,
-  canvasPoolManager: CanvasPoolManager
+  canvasPoolManager: CanvasPoolManager,
+  disableZoom: boolean = false
 ) {
   const zoom = useCallback(
     (evt: WheelEvent) => {
-      if (!canvas || !view) {
+      if (!canvas || !view || disableZoom) {
```
[the remark this answers is not in the document]
Comment (eliashussary) [reply static/app/components/profiling/flamegraph/interactions/useWheelCenterZoom.tsx:17]: ✅

File: static/app/components/profiling/flamegraph/interactions/useWheelCenterZoom.tsx [hunk 6]
```diff
@@ -28,7 +29,7 @@ export function useWheelCenterZoom(
         view,
       ]);
     },
-    [canvas, view, canvasPoolManager]
+    [canvas, view, canvasPoolManager, disableZoom]
```
[the remark this answers is not in the document]
Comment (eliashussary) [reply static/app/components/profiling/flamegraph/interactions/useWheelCenterZoom.tsx:32]: ✅

File: static/app/components/profiling/flamegraph/interactions/useWheelCenterZoom.tsx [hunk 7]
```diff
@@ -9,11 +9,12 @@ import {getCenterScaleMatrixFromMousePosition} from 'sentry/utils/profiling/gl/u
 export function useWheelCenterZoom(
   canvas: FlamegraphCanvas | null,
   view: CanvasView<any> | null,
-  canvasPoolManager: CanvasPoolManager
+  canvasPoolManager: CanvasPoolManager,
+  disableZoom: boolean = false
```
[the remark this answers is not in the document]
Comment (eliashussary) [reply static/app/components/profiling/flamegraph/interactions/useWheelCenterZoom.tsx:13]: ✅

File: static/app/components/profiling/flamegraph/aggregateFlamegraph.tsx [hunk 8]
```diff
@@ -0,0 +1,415 @@
...
+    [flamegraph, flamegraphCanvas, flamegraphTheme]
+  );
+
+  useEffect(() => {
+    const canvasHeight = flamegraphCanvasRef?.getBoundingClientRect().height;
+    if (!canvasHeight) {
+      return undefined;
+    }
+
+    const removeModifier = setFlamegraphThemeMutation(theme => {
```
[the remark this answers is not in the document]
Comment (eliashussary) [reply static/app/components/profiling/flamegraph/aggregateFlamegraph.tsx:210]: to be frank, we don't really need this. 

i had another version of this where we pushed and popped modifiers to a set, but it was overkill and simply not needed. i just need to clean this up.

it would only ever be useful if we're rendering `ThemeProvider` once, and we need to mutate state and in transition we need to remove the mutation for the next view etc.

tldr; useless over optimization based on the final version of my code, i'll rm it

File: static/app/components/profiling/flamegraph/aggregateFlamegraph.tsx (hunk 1 above)
Comment (eliashussary) [reply static/app/components/profiling/flamegraph/aggregateFlamegraph.tsx:63]: ✅

File: static/app/utils/profiling/flamegraph/flamegraphThemeProvider.tsx (hunk 4 above)
Comment (JonasBa) [reply static/app/utils/profiling/flamegraph/flamegraphThemeProvider.tsx:14]: >Absolutely we should be passing via props.

We did this at the start and it was just a lot of prop drilling hence why we removed it, but at least for each individual renderer, we should only be passing what we need. I have not foreseen a case where we'd want to make this dynamic (not from like a child component pov like here), but hindsight is 20/20 or something - we should keep this in mind as we are imo approaching a time when we might want to improve this part of the codebase

**Review by JonasBa — APPROVED** [review]
